### PR TITLE
Cleaning mypy errors in utils, type_utils, udf_utils, and telemetry.

### PR DESCRIPTION
**Pull request by sfc-gh-jrose**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

This PR cleans up mypy errors in the following files:
```
src/snowflake/snowpark/_internal/telemetry.py 
src/snowflake/snowpark/_internal/type_utils.py 
src/snowflake/snowpark/_internal/udf_utils.py 
src/snowflake/snowpark/_internal/utils.py
```

### CODE REVIEW

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 1]
```diff
@@ -22,7 +22,7 @@
     get_type_hints,
 )
 
-import cloudpickle
+import cloudpickle  # type: ignore
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/_internal/udf_utils.py:25]: Cloudpickle lacked type information so ignoring it so mypy is queit.

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 2]
```diff
@@ -186,95 +186,120 @@ def extract_return_type_from_udtf_type_hints(
...
+    elif object_type in (TempObjectType.FUNCTION, TempObjectType.PROCEDURE):
+        python_types_dict = retrieve_func_type_hints_from_source(filename, func_name)
+    else:
+        raise ValueError(
+            f"Expecting FUNCTION, PROCEDURE, TABLE_FUNCTION, or AGGREGATE_FUNCTION as object_type, got {object_type}"
+        )
+    return python_types_dict or {}
+
+
+def get_types_from_type_hints(
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/_internal/udf_utils.py:285]: This function was overly verbose and reused variable names even though they had slightly different meanings. Separating out the differing logic makes it a little more readable and clears up some mypy errors.

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 3]
```diff
@@ -446,7 +478,7 @@ def extract_return_input_types(
             return_type_from_type_hints, PandasDataFrameType
         ):  # vectorized UDTF
             return_type = PandasDataFrameType(
-                [x.datatype for x in return_type], [x.name for x in return_type]
+                [x.datatype for x in return_type], [x.name for x in return_type]  # type: ignore
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/_internal/udf_utils.py:481]: There's some sort of coupling between the `return_type_from_type_hints` and `return_type` variables that mypy isn't able to understand. Fixing would require a larger refactor than I think is appropriate for this PR.

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 4]
```diff
@@ -499,7 +531,7 @@ def extract_return_input_types(
                 True,
                 False,
                 res_return_type.element_type,
-                [tp.element_type for tp in res_input_types],
+                [tp.element_type for tp in res_input_types],  # type: ignore
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/_internal/udf_utils.py:534]: All the elements in this list have been validated to be `PandasSeriesType`, but mypy isn't picking up on that.

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 5]
```diff
@@ -1124,7 +1164,7 @@ def generate_call_python_sp_sql(
     sql_args = []
     for arg in args:
         if isinstance(arg, snowflake.snowpark.Column):
-            sql_args.append(session._analyzer.analyze(arg._expression, {}))
+            sql_args.append(session._analyzer.analyze(arg._expression, {}))  # type: ignore
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/_internal/udf_utils.py:1167]: mypy things the `_analyzer` object doesn't have an analyze method, but both Analyzer and MockAnalyzer types have the analyze method.

File: src/snowflake/snowpark/_internal/udf_utils.py (hunk 1 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/_internal/udf_utils.py:25]: We could use something like:
```
[mypy-cloudpickle.*]
ignore_missing_imports = True
```
in `mypy.ini`

File: src/snowflake/snowpark/_internal/utils.py [hunk 6]
```diff
@@ -49,7 +49,9 @@
     try:
         from snowflake.connector.cursor import ResultMetadataV2
     except ImportError:
-        ResultMetadataV2 = ResultMetadata
+        from snowflake.connector.cursor import (  # type: ignore
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/utils.py:52]: Can we just not import here? Looks like `ResultMetadataV2` is only used for type checking.

File: src/snowflake/snowpark/_internal/udf_utils.py (hunk 2 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/_internal/udf_utils.py:285]: This is great, thanks for refactoring this!

**Comment by sfc-gh-stan** [comment]
Let's use use `# type: ignore[<code>]` to only ignore specific errors.

**Review by sfc-gh-stan — APPROVED** [review]

**Pull request merged**